### DISCUSSION TITLE
Multi-app System Run Sub App Every Other Time Step

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#26529] [Q&A General]
What would be the best way to be able to run sub-apps every nth timestep?

**Comment by GiudGiud** [comment]
I would use Controls, the conditional enable ones
https://mooseframework.inl.gov/syntax/Controls/index.html

I d probably bake the nth timestep logic into a ParsedFunction or a ParsedPostprocessor (which can be used in a ParsedFunction)

**Reply by ghost** [reply]
So would I set the enabled flag on the TransientMultiApp to be equal to the value of a ParsedFunctionControl block?

**Reply by GiudGiud** [reply]
unfortunately that wont work, the ParsedFunctionControl is for THM simulations and we dont use it for enabling objects.

I would use a ParsedFunction inside a https://mooseframework.inl.gov/source/controls/ConditionalFunctionEnableControl.html

**Reply by ghost** [reply]
Thanks. This seems to mostly work. It does the enable and disable, but it does not seem to work on the correct initial time step. Below is what I added...

```
[AuxVariables]
  [timesteps]
    family = SCALAR
    initial_condition = 0  
  []
[]

[AuxScalarKernels]
  [count_steps]
    type = FunctionScalarAux
    variable = timesteps
    execute_on = TIMESTEP_BEGIN
    function = fcount_steps
  []
[]

[Functions]
  [fcount_steps]
    type = ParsedFunction
    value = 'a+1'
    vars = 'a'
    vals = 'timesteps'
  []
  [subapp_ctrl]
    type = ParsedFunction
    value = 'a%5=0'
    vars = 'a'
    vals = 'timesteps'
  []
[]

[Controls]
  [disable]
      type = ConditionalFunctionEnableControl
      conditional_function = subapp_ctrl
      enable_objects = 'MultiApps::sub'
      execute_on = TIMESTEP_END 
  []
[]
```

What I was trying to have it do was have the subapp run every 5th timestep and it does on every one but the first run. Right now the subapp will run on the 8th, 13th and 18th timesteps. I can see the scalar outputs of the timestep variable increments correctly with time, but it appears that the `ConditionalFunctionEnableControl` is out of sync with the aux system updating the timesteps variable.

The 5th timestep isn't a critical number, it was just a value that would run often enough in the test while also clearly showing a break between runs in the log output.

Thanks!

**Reply by GiudGiud** [reply]
Hello

I could see that. There's sort of an sequence of execution of objects before the updated value gets seen, 3 steps is a little excessive here. I would have expected maybe 1 or 2.

I would say we need a NumTimeSteps postprocessor. We dont have that. So let's use these two:
https://mooseframework.inl.gov/source/materials/TimeStepMaterial.html that provides a timestep material property
https://mooseframework.inl.gov/source/postprocessors/ElementExtremeMaterialProperty.html
then feed the PP in the ParsedFunction. 

If you execute the postprocessor often enough, it will be up to date for the evaluation of the function done on step 5.

Guillaume

**Reply by ghost** [reply]
That method seems a bit cleaner to get the timesteps, but I can't seem to get the PP to execute often enough. If I set `execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END'` it now starts on step 7 and goes every 5 after that. But if I do `execute_on = 'ALWAYS'` the PP doesn't seem to update at all.

```
Postprocessor Values:
+----------------+----------------+----------------+----------------+
| time           | gap_distance   | gap_mm         | timestep_pp    |
+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   3.000000e-04 |   3.000000e-01 |   0.000000e+00 |
|   2.000000e+00 |   3.000000e-04 |   3.000000e-01 |   0.000000e+00 |
|   3.000000e+00 |   9.718305e-05 |   9.718305e-02 |   0.000000e+00 |
|   4.000000e+00 |   1.025832e-04 |   1.025832e-01 |   0.000000e+00 |
|   5.000000e+00 |   1.025832e-04 |   1.025832e-01 |   0.000000e+00 |
+----------------+----------------+----------------+----------------+
```

This is how I setup the material and PP
```
[timestep_pp]
      type = ElementExtremeMaterialProperty
      mat_prop = time_step
      value_type = max
      # execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END'
      execute_on = 'ALWAYS'
[]
[time]
    type = TimeStepMaterial
[]
```

I think a NumTimeSteps PostProcessor would be good to have, but I am thinking it will not solve the issue. I would still need to adjust the ParsedFunction to trigger earlier to account for the delay.

**Reply by GiudGiud** [reply]
ALWAYS should no longer exist for postprocessors. Please do not use that, it was removed for a reason (not as often as always would suggest)

Step 7 is better I guess. I think I would understand 6. Since you start at 1 at the end of timestep 1, then I could see 5 at the end of timestep 5 being too late for the Control (which does not execute at the very end)

Would you be open to just doing this?
```
  [subapp_ctrl]
    type = ParsedFunction
    value = '(a-2)%5=0'
    vars = 'a'
    vals = 'timesteps'
  []
[]
```
I know it s not super clean, but it should trigger on 5

Another thing would be to execute the control on timestep begin.

**Reply by ghost** [reply]
Ah ok. 

Yeah, making the change like that shouldn't be an issue. Do you think it would be consistent with the offset? Or if I changed the number of steps to skip would I also need to change the `a-2` to a different offset?

**Reply by GiudGiud** [reply]
I would expect the offset and the frequency (currently 5) to be independent.

**Reply by ghost** [reply]
Ok thanks! 

I will move forward with the PP approach since it requires less additions to the input file and update the offset.

Appreciate the help.
